onebot/plugins/karma.py

```diff
@@ -0,0 +1,81 @@
+# -*- coding: utf-8 -*-
+"""
+================================================
+:mod:`onebot.plugins.karma` Karma
+================================================
+
+This plugin keeps track of karma.
+
+To modify: foo++ / (foo bar)++ / foo--
+To view: !karma foo bar
+"""
+
+import base64
+import irc3
+from irc3 import plugin
+from irc3.plugins.command import command
+
+
+@plugin
+class KarmaPlugin(object):
+    """Karma Plugin"""
+
+    requires = [
+        'irc3.plugins.storage',
+        'irc3.plugins.command',
+    ]
+
+    def __init__(self, bot):
+        self.bot = bot
+
+    @staticmethod
+    def _find_key(data, modifier='++'):
+        # This assumes that `modifier' is present in `data'
+        data = data[:data.find(modifier)]
+        if data[-1] == ')':
+            key = data.split('(')[-1][:-1]
+        else:
+            key = data.split(' ')[-1]
+        return KarmaPlugin._format_key(key)
+
+    @staticmethod
+    def _format_key(key):
+        return 'karma_{}'.format(base64.b64encode(key.lower().encode('utf-8')))
+
+    def _update_karmadict(self, karmakey, field):
+        karma = {'up': 0, 'down': 0}
+        if karmakey in self.bot.db:
+            karma = self.bot.db[karmakey]
+        karma[field] += 1
+        self.bot.db[karmakey] = karma
+
+    @irc3.event((r'''(@(?P<tags>\S+) )?:(?P<mask>\S+) (PRIVMSG)'''
+                 r''' (?P<target>\S+)(\s:(?P<data>.+\+\+)|$)'''))
```

Comment:
how does this handle multiple occurances per line? e.g.

    <Thom> foo++ bar++